frame/contracts/fixtures/float_instruction.wat

```diff
@@ -1,5 +1,6 @@
 ;; Module that contains a float instruction which is illegal in deterministic mode
 (module
+	(import "env" "memory" (memory 1 1))
```

Comment:
> do you mean that there could be some contracts having no memory imports uploaded on chain, which would not pass this check made during the instantiation?

Yes. It will break existing contracts.

> This was done as a workaround to having the [`ensure_no_internal_memory()`](https://github.com/paritytech/substrate/blob/master/frame/contracts/src/wasm/prepare.rs#L92C28-L92C28) check, which seems impossible to be accomplished with wasmi API.

Can't you [iterate over the imports](https://docs.rs/wasmi/0.30.0/wasmi/struct.Module.html#method.imports) and reject if you find a memory?

Reply:
> Can't you [iterate over the imports](https://docs.rs/wasmi/0.30.0/wasmi/struct.Module.html#method.imports) and reject if you find a memory?

We do iterate over the imports in [scan_imports()](https://github.com/paritytech/substrate/blob/56dc2dfe8cc0ed90023242d6fd2bf5b63186551e/frame/contracts/src/wasm/prepare.rs#L176), still sorry I don't get what you mean, why to reject memory import? We do [reject](https://github.com/paritytech/substrate/blob/56dc2dfe8cc0ed90023242d6fd2bf5b63186551e/frame/contracts/src/wasm/prepare.rs#L210) a second memory import, but striclty speaking this would not replace the aforementioned [`ensure_no_internal_memory()`](https://github.com/paritytech/substrate/blob/master/frame/contracts/src/wasm/prepare.rs#L92C28-L92C28) coz what it catches is internal mem declarations like e.g. this one:

```wat
(module
 	(memory 1 1)

	(func (export "deploy"))
	(func (export "call"))
)
```

Reply:
Yes sure my brainfart. We can't check for internal memory. As far as I see it the best way would be as you suggested: Only enforce the check for newly uploaded code.

Reply:
> Only enforce the check for newly uploaded code.

implemented it this way:
- Require module to have a memory import when initially validated (`prepare::validate()`), which only happens during code upload.
- Omit this check and fallback to (0,0) when instantiating the module for engine execution ( in `WasmBlob::instantiate()`).
  falling back to (0,0) is the logic we already [had there](https://github.com/paritytech/substrate/blob/37e14070acbf432e6c82857d9fc7cd4dbd7fa303/frame/contracts/src/wasm/prepare.rs#L335) for a while

Reply:
> Yes. It will break existing contracts.

However, some research shows that currently no contract without memory import is deployed on parachains.
Basically cargo-contract [adds](https://github.com/paritytech/cargo-contract/blob/master/crates/build/src/args.rs#L196C81-L196C95) `--import-memory` flag to the lld so every contract built with it has the import. 

@athei given that facts, are you still for omiiting the check for already stored contracts?

Reply:
> Only enforce the check for newly uploaded code.

So as has been agreed in the channel discussion, [rolled back](https://github.com/paritytech/substrate/blob/bf5277062d4460273a8ba27badad18f0781d3489/frame/contracts/src/wasm/mod.rs#L228) this change. Now we **always** enforce this check for **all** modules. 

Only hand-written .wat contracts currently (before we merge this PR) could have no memory imports. We haven't found any of those ones deployed on production perachains currently. And

> No useful interaction with the chain is possible without memory anyways.